std/experimental/checkedint.d

```diff
@@ -562,6 +562,44 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
         }
     }
 
+    /// ditto
+    size_t toHash(this This)() const nothrow @safe if (is(This == shared))
```

Comment:
This signature is because when I used

```d
size_t toHash()() const shared nothrow @safe
```
I got this error:
>Error: `std.experimental.checkedint.Checked!(int, Hook1).Checked.toHash` called with argument types `() immutable` matches both:
>`std.experimental.checkedint.Checked!(int, Hook1).Checked.toHash() const`
>and:
>`std.experimental.checkedint.Checked!(int, Hook1).Checked.toHash!().toHash() shared const`

I don't think that should have been an error since non-template functions are supposed to always match before template functions.

Reply:
Wouldn't it be better if you merged both toHash functions into a single templated this one and avoid the code duplication?

Reply:
@RazvanN7 I need to retain a non-template version of `toHash` or else `TypeInfo.getHash` won't pick it up. Demonstration:

https://run.dlang.io/is/gyqq5g
```d
import std.stdio: writeln;

struct S1
{
    size_t toHash() const nothrow @safe { return 44; }
}

struct S2
{
    size_t toHash(this _)() const nothrow @safe { return 44; }
}

void main()
{
    S1 s1;
    S2 s2;
    writeln(s1.toHash()); // 44
    writeln(typeid(s1).getHash(&s1)); // 44
    writeln(s2.toHash()); // 44
    writeln(typeid(s2).getHash(&s2)); // 1364076727
}
```

Reply:
Darn... OK, this is good then.